### ISSUE TITLE
Add a hook for when a function returns

### ISSUE THREAD

**Issue by maxnordlund** [fish-shell/fish-shell#2828]
Moved from #2036:

---

I not sure if this belongs here, but I can't seem to find a way to trap function return. That is, do some cleanup when a function returns, regardless if it was interrupted somewhere in the middle.

```fish
function foo
  set -l dir (mktemp -d temporary_stuff)
  function cleanup --on-function-return --inherit-variable dir
    rm -rf $dir
    function --erase cleanup # Should be done automatically?
  end

  # stuff with foo
end
```

---

@maxnordlund: There is no --on-function-return option for the function command. I'd recommend opening an enhancement issue as I couldn't find an open issue asking for such a feature and it's orthogonal to this issue. It's also unclear whether your proposal is the best way to solve this issue. For example would a "try; ...; except" style block be better and solve other problems? Or would it be better to add define foo like function foo --on-return=func_name and have the named function automatically inherit the scope of "foo"?

---

Yes I know it doesn't exist, but wasn't sure if it was different enough to warrant it's own issue, as I belive it is not orthogonal, rather slightly angled to this one.

Using `try`/`catch`/`finally` would not be the quite same, and would mean that every function that needs some cleanup needs to wrap its entire body in an `try` statement. It also confuses events/signals with exceptions, which may or may not be such a good idea.

The extra flag to `function` does sound like a nicer alternative.

**Comment by ridiculousfish** [comment]
This sort of defer-style construct seems pretty heavyweight for a shell scripting language. The "do something even if interrupted" is also a little scary, since normally the user interrupts to get control back.

**Comment by maxnordlund** [comment]
Yeah, but I guess you can be more specific if the `--on-signal` flag worked better. What I'm try to get at is the ability to unconditionally perform some cleanup when the execution of a function is done.

Essentially a traditional `finally`, but for the whole function. In bash you can just use `trap` and call the cleanup function manually at the end in case the user didn't interrupt the main function.

About doing things when interrupted, yeah that can be scary, but that's what `trap`/`--on-signal` is for, right?

**Comment by alphapapa** [comment]
Yeah, AFAIK, trapping to cleanup when interrupted is pretty standard in shell scripts, e.g.:

```sh
#/bin/bash

cleanup() {
    rm -f $tempfile
}

tempfile=$(mktemp)

trap cleanup INT TERM
```

You can play with trapping with this Bash script:
```sh
#!/bin/bash

cleanup() {
    sleep 2
    echo cleanup done
}

trap cleanup INT TERM

sleep 5
echo slept once

sleep 5
echo slept twice

echo script finished
```

You may have noticed having to mash Ctrl+C repeatedly in some random script to get back to a prompt.  That illustrates the race condition between starting the next command in the script, which steals the signal from the script's process; if you mash it fast enough or hold it down, the shell gets the signal first and exits the script.

About this feature, I'm not sure if I understand why it's necessary or how it would interact with trapping.  Wouldn't an `--on-function-return` option be essentially the same as, e.g.:

```fish
function something
    sleep 5
    echo something
end

something
after-something
```

`after-something` would run if `something` were interrupted or if it finished.

**Comment by krader1961** [comment]
> Wouldn't an --on-function-return option be essentially the same as, e.g.: ....

No, because the intent is for the caller to be unaware that some cleanup actions might be needed. If I write a `nifty_cool_function` and you want to use it you shouldn't have to know that my function creates a temporary file which should be removed regardless of how the function returns. In Python you would normally use the `with` statement or a `try...finally...` block for this purpose.

**Comment by maxnordlund** [comment]
For a real world example of a complex function, and the reason I asked about this I have an [update](https://github.com/maxnordlund/dotfiles/blob/master/fish/functions/update.fish) function that uses temporary files to keep state between the phases of the updates.

**Comment by alphapapa** [comment]
> No, because the intent is for the caller to be unaware that some cleanup actions might be needed. If I write a nifty_cool_function and you want to use it you shouldn't have to know that my function creates a temporary file which should be removed regardless of how the function returns.

That's an interesting idea, although I'm not sure how I feel about that concept in shell scripting.  Couldn't you write the function as a separate script and call it as a command?

**Comment by krader1961** [comment]
> Couldn't you write the function as a separate script and call it as a command?

In some cases, yes, but not if the function needs to modify the state of the current shell. And an external command is always more expensive.

**Comment by alphapapa** [comment]
The idea of a function with hidden side effects modifying the state of the current shell shivers me timbers.  :)

**Comment by krader1961** [comment]
> The idea of a function with hidden side effects modifying the state of the current shell shivers me timbers.

It shouldn't. A lot of functions do so; consider the `cd` function that maintains a history of recently visited directories. There are basically two reasons to use a function rather than external script; 1) speed, 2) to have access to the state of the current shell.

**Comment by zanchey** [comment]
The psub function does something similar, I think:

```
        # Find unique function name
        while true
                set funcname __fish_psub_(random);
                if not functions $funcname >/dev/null ^/dev/null
                        break;
                end
        end

        # Make sure we erase file when caller exits
        function $funcname --on-job-exit caller --inherit-variable filename --inherit-variable dirname --inherit-variable funcname
                command rm $filename
                if count $dirname >/dev/null
                        command rmdir $dirname
                end
                functions -e $funcname
        end
```

In particular, is `--on-job-exit caller` enough?

**Comment by maxnordlund** [comment]
Unfortunately no, because it can't be used for a function running in the foreground. But if it were possible to tweak it and allow something simliar for any function, then yes it would solve the problem. That is what I propose here.

**Comment by clarfonthey** [comment]
I also ran into this today when trying to work with temp files. I personally would be amenable to just having a fish builtin which could create temporary files and delete them whenever a function returns, but I feel like this would be more complicated than simply providing an internal way to defer these.

Personally, what makes the most sense to me is a "defer to scope" builtin, which is most akin to `try..finally` blocks in other languages. Function returns would be just a special case of this. The main benefit of this is stuff like temporary files, or other resource cleanup when temporary resources need to be created.

**Comment by plax-00** [comment]
Would love to see this. In zsh I was able to `trap cleanup EXIT` to get this behavior but in fish this only triggers when the shell itself exits.
